services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/FlowDiversitySpec.groovy

```diff
@@ -0,0 +1,417 @@
+            }.size() == broughtDownPorts.size() * 2
+        }
+
+        when: "Create the second flow with diversity enabled"
+        def flow2 = flowHelper.randomFlow(srcSwitch, dstSwitch, false, [flow1]).tap { it.diverseFlowId = flow1.id }
+        flowHelper.addFlow(flow2)
+        def flow2Path = PathHelper.convert(northbound.getFlowPath(flow2.id))
+
+        then: "The second flow is built through the same path as the first flow"
+        flow2Path == flow1Path
+
+        and: "Restore topology, delete flows and reset costs"
+        broughtDownPorts.every { northbound.portUp(it.switchId, it.portNo) }
+        [flow1, flow2].each { flowHelper.deleteFlow(it.id) }
+        Wrappers.wait(discoveryInterval + WAIT_OFFSET) {
+            northbound.getAllLinks().each { assert it.state != IslChangeType.FAILED }
+        }
+        database.resetCosts()
+    }
+
+    def "Links and switches get extra cost that is considered when calculating diverse flow paths"() {
+        given: "Two active neighboring switches with three not overlapping paths at least"
+        def (Switch srcSwitch, Switch dstSwitch) = getSwitchPair(3)
+
+        and: "Create a flow going through these switches"
+        def flow1 = flowHelper.randomFlow(srcSwitch, dstSwitch)
+        flowHelper.addFlow(flow1)
+        def flow1Path = PathHelper.convert(northbound.getFlowPath(flow1.id))
+
+        and: "Make all min alternative paths less preferable by 1 for diverse flows"
+        def altPaths = database.getPaths(srcSwitch.dpId, dstSwitch.dpId)*.path
+        altPaths.remove(flow1Path)
+        def flow1PathCost = pathHelper.getCost(flow1Path) + diversityIslWeight + diversitySwitchWeight * 2
+        def minAltPathCost = pathHelper.getCost(altPaths.min { it.size() }) + diversitySwitchWeight * 2
+        if (flow1PathCost > minAltPathCost) {
+            int minAltPathNodeCount = altPaths.min { it.size() }.size()
+            int difference = flow1PathCost - minAltPathCost
+            altPaths.findAll { it.size() == minAltPathNodeCount }.each {
+                def isl = pathHelper.getInvolvedIsls(it)[0]
+                int islCost = database.getIslCost(isl)
+                northbound.updateLinkProps([new LinkPropsDto(isl.srcSwitch.dpId.toString(), isl.srcPort,
+                        isl.dstSwitch.dpId.toString(), isl.dstPort, ["cost": (islCost + difference + 1).toString()])])
+            }
+        }
+
+        when: "Create the second flow with diversity enabled"
+        def flow2 = flowHelper.randomFlow(srcSwitch, dstSwitch, false, [flow1]).tap { it.diverseFlowId = flow1.id }
+        flowHelper.addFlow(flow2)
+        def flow2Path = PathHelper.convert(northbound.getFlowPath(flow2.id))
+
+        then: "The flow is built through the most preferable path (path of the first flow)"
+        flow2Path == flow1Path
+
+        when: "Create the third flow with diversity enabled"
+        if (flow1PathCost < minAltPathCost) {
```

Comment:
One more `if`, which is actually a wrong assumption from my point of view, since `minAltPathCost` has already changed on lines L180-L189(maybe changed or maybe not ¯\\_(ツ)_/¯), so it is operating with potentially outdated values. or is it intentional?
`flow1PathCost` is also outdated at this point, sinse flow1PathCost now has the `flow2` on it, so the 'diverse cost' of this path has changed as soon as you built a second flow through it